sarracen/interpolate.py

```diff
@@ -302,3 +302,120 @@ def _fast_interpolate3d_cross(xterm, yterm, zterm, wfunc, zslice, kernrad, targe
+                                data['h'].to_numpy(),
+                                xmin,
+                                ymin,
+                                pixwidthx,
+                                pixwidthy,
+                                pixcountx,
+                                pixcounty)
+
+
+# Underlying numba-compiled code for 3D column interpolation.
+@numba.jit(nopython=True, parallel=True, fastmath=True)
+def _fast_3d_interpolate(xparts, yparts, target, wfuncint, int_samples, wrad, mass, rho, h, xmin, ymin,
+                         pixwidthx, pixwidthy, pixcountx, pixcounty):
+    image = np.zeros((pixcounty, pixcountx))
+
+    term = target * mass / (rho * h ** 2)
+
+    # determine maximum and minimum pixels that each particle contributes to
+    ipixmin = np.rint((xparts - wrad * h - xmin) / pixwidthx) \
+        .clip(a_min=0, a_max=pixcountx)
+    jpixmin = np.rint((yparts - wrad * h - ymin) / pixwidthy) \
+        .clip(a_min=0, a_max=pixcounty)
+    ipixmax = np.rint((xparts + wrad * h - xmin) / pixwidthx) \
+        .clip(a_min=0, a_max=pixcountx)
+    jpixmax = np.rint((yparts + wrad * h - ymin) / pixwidthy) \
+        .clip(a_min=0, a_max=pixcounty)
+
+    # iterate through the indexes of non-filtered particles
+    for i in prange(len(term)):
+        # precalculate differences in the x-direction (optimization)
+        dx2i = ((xmin + (np.arange(int(ipixmin[i]), int(ipixmax[i])) + 0.5) * pixwidthx - xparts[i]) ** 2) \
+               * (1 / (h[i] ** 2))
+
+        # determine differences in the y-direction
+        ypix = ymin + (np.arange(int(jpixmin[i]), int(jpixmax[i])) + 0.5) * pixwidthy
+        dy = ypix - yparts[i]
+        dy2 = dy * dy * (1 / (h[i] ** 2))
+
+        # calculate contributions at pixels i, j due to particle at x, y
+        q2 = dx2i + dy2.reshape(len(dy2), 1)
+        wab = np.interp(np.sqrt(q2), np.linspace(0, wrad, int_samples), wfuncint)
+
+        # add contributions to image
+        image[int(jpixmin[i]):int(jpixmax[i]), int(ipixmin[i]):int(ipixmax[i])] += (wab * term[i])
+
+    return image
+
+
+def _get_column_kernel(kernel, samples):
+    """
+    Generate a 2D column kernel approximation, by integrating a given 3D kernel over the z-axis.
+    :param kernel: The 3D kernel to integrate over.
+    :param samples: The number of samples to take of the integral.
+    :return: A ndarray of length (samples), containing the kernel approximation.
+    """
+    results = []
+    for sample in np.linspace(0, kernel.get_radius(), samples):
+        results.append(2 * quad(_int_func,
+                                a=0,
+                                b=np.sqrt(kernel.get_radius() ** 2 - sample ** 2),
+                                args=(sample, kernel))[0])
+
+    return np.array(results)
+
+
+# Internal function for performing the integral in _get_column_kernel()
+def _int_func(q, a, kernel):
+    return kernel.w(np.sqrt(q ** 2 + a ** 2), 3)
```

Comment:
I go back and forth on whether column integration of the kernel should be part of the kernel instead of here. What do you think? We can discuss.